pubtools/_impl/tracing.py

```diff
@@ -0,0 +1,138 @@
+"""Instrument functions.
+
+Usage:
+    @instrument_func()
+      def func():
+          pass
+
+"""
+import functools
+import logging
+import os
+import threading
+
+from opentelemetry import baggage, context, trace
+from opentelemetry.baggage.propagation import W3CBaggagePropagator
+from opentelemetry.propagate import set_global_textmap
+from opentelemetry.sdk.resources import SERVICE_NAME, Resource
+from opentelemetry.sdk.trace import TracerProvider
+from opentelemetry.sdk.trace.export import (BatchSpanProcessor,
+                                            ConsoleSpanExporter)
+from opentelemetry.trace import Status, StatusCode
+from opentelemetry.trace.propagation.tracecontext import \
+    TraceContextTextMapPropagator
+
+from pubtools.pluggy import pm
+
+propagator = TraceContextTextMapPropagator()
+baggage_propagator = W3CBaggagePropagator()
+log = logging.getLogger(__name__)
+
+
+def get_trace_wrapper():
+    """return a global trace wrapper instance"""
+    return TracingWrapper()
```

Comment:
There is no special case here when tracing is disabled.  Isn't it going to lead to a crash?  You've implemented `__new__` so it'll return None in that case.

Then won't the following happen?

```python
from pubtools.tracing import get_trace_wrapper

tw = get_trace_wrapper()

@tw.instrument_func(...)  # CRASH!  When tracing is disabled, tw is None so this can't work.
def some_func(...):
    ....
```

That's why I said it'd have to return an object which "does nothing" (i.e. `instrument_func` exists but has no effect) in the case of tracing being disabled.

This is not necessarily just about instrument_func either.  You said that you expect downstream code to access tw.processor, that would also crash if tw is None.  Will everywhere accessing tw.processor need to first check if tw is None, or maybe check if tw.processor is None?  It would be nice to come up with some better way to handle it because that will lead to bugs where people write code which only works when tracing is enabled, or vice-versa.

One option would be not to expose `processor` at all, but instead on the TracingWrapper class you implement the specific methods that you expect downstream to need (e.g. have a `TracingWrapper.flush()` method and have it do nothing if tracing is disabled).

By the way, I'm not familiar with the otel python library at all, but it wouldn't surprise me if it already has some way to make it gracefully do nothing.  If so it would be cleaner to use that than to have to introduce None vs non-None code paths.

Reply:
It did crash. I will fix it.